src/compas/geometry/surfaces/nurbs.py

```diff
@@ -0,0 +1,381 @@
+from itertools import product
+
+from compas.geometry import Point
+from compas.utilities import meshgrid, linspace
+
+from ._surface import Surface
+
+
+class NurbsSurface(Surface):
+    """Class representing a NURBS surface.
+
+    Attributes
+    ----------
+    points: List[Point]
+        The control points of the surface.
+    weights: List[float]
+        The weights of the control points.
+    u_knots: List[float]
+        The knot vector, in the U direction, without duplicates.
+    v_knots: List[float]
+        The knot vector, in the V direction, without duplicates.
+    u_mults: List[int]
+        The multiplicities of the knots in the knot vector of the U direction.
+    v_mults: List[int]
+        The multiplicities of the knots in the knot vector of the V direction.
+    u_degree: int
+        The degree of the polynomials in the U direction.
+    v_degree: int
+        The degree of the polynomials in the V direction.
+    u_domain: Tuple[float, float]
+        The parameter domain in the U direction.
+    v_domain: Tuple[float, float]
+        The parameter domain in the V direction.
+    is_u_periodic: bool
+        True if the curve is periodic in the U direction.
+    is_v_periodic: bool
+        True if the curve is periodic in the V direction.
+
+    """
+
+    @property
+    def DATASCHEMA(self):
+        from schema import Schema
+        from compas.data import is_float3
+        from compas.data import is_sequence_of_int
+        from compas.data import is_sequence_of_float
+        return Schema({
+            'points': lambda points: all(is_float3(point) for point in points),
+            'weights': is_sequence_of_float,
+            'u_knots': is_sequence_of_float,
+            'v_knots': is_sequence_of_float,
+            'u_mults': is_sequence_of_int,
+            'v_mults': is_sequence_of_int,
+            'u_degree': int,
+            'v_degree': int,
+            'is_u_periodic': bool,
+            'is_v_periodic': bool
+        })
+
+    @property
+    def JSONSCHEMANAME(self):
+        raise NotImplementedError
+
+    def __init__(self, name=None):
+        super().__init__(name=name)
```

Comment:
Again, ipy-friendliness? along with all the numpy/lru_cache stuff that follows...